### PR TITLE
Allow configuring a relationship to an entity type without a primary key

### LINKED ISSUE #1603: Configuration: defining a relationship to an entity type without primary key shouldn't throw
It should be possible to build a model like this:
```
modelBuilder.Entity<CustomerDetails>()
    .HasOne(e => e.Customer)
    .WithOne(e => e.Details)
    .ReferencedKey<Customer>(e => e.Name);
modelBuilder.Entity<Customer>()
    .Key(e => e.Name);
```

### PR DESCRIPTION
**Pull request by AndriySvyryd**
Remove shadow keys added by convention when the primary key is set
Update relationships when a new primary key is set

Fixes #1603

### CODE REVIEW

File: src/EntityFramework.Core/Metadata/Internal/InternalEntityBuilder.cs [hunk 1]
```diff
@@ -52,27 +52,72 @@ private InternalKeyBuilder PrimaryKey(IReadOnlyList<Property> properties, Config
...
+            ReplaceConventionShadowKeys(keyBuilder.Metadata);
+
+            return keyBuilder;
+        }
+
+        private void UpdateReferencingForeignKeys(Key keyToReplace, Key newKey, ConfigurationSource configurationSource)
+        {
+            var newProperties = newKey.Properties;
+
+            var allForeignKeysReplaced = true;
```
Comment (anpete) [inline src/EntityFramework.Core/Metadata/Internal/InternalEntityBuilder.cs]: This looks like LINQ Aggregate.

File: src/EntityFramework.Core/Metadata/Internal/InternalEntityBuilder.cs [hunk 2]
```diff
@@ -997,6 +1016,17 @@ private InternalIndexBuilder Index(IReadOnlyList<Property> properties, Configura
             {
                 var baseName = (string.IsNullOrEmpty(navigationToPrincipal) ? principalType.SimpleName : navigationToPrincipal);
 
+                if (principalKey == null)
+                {
+                    var principalKeyProperty = CreateUniqueProperty(
+                        "Id",
```
Comment (anpete) [inline src/EntityFramework.Core/Metadata/Internal/InternalEntityBuilder.cs]: Should this be a constant?

File: src/EntityFramework.Core/Metadata/Property.cs [hunk 3]
```diff
@@ -210,6 +210,46 @@ private void SetFlag(bool? value, PropertyFlags flag)
 
         internal static string Format(IEnumerable<IProperty> properties) => "{" + string.Join(", ", properties.Select(p => "'" + p.Name + "'")) + "}";
 
+        public static bool AreCompatible([NotNull] IReadOnlyList<Property> principalProperties,
+            [NotNull] IReadOnlyList<Property> dependentProperties)
+        {
+            return IsPropertyCountSame(principalProperties, dependentProperties)
```
Comment (anpete) [inline src/EntityFramework.Core/Metadata/Property.cs]: Seems over factored.

**Comment by anpete** [comment]
:shipit:

File: src/EntityFramework.Core/Metadata/Internal/InternalEntityBuilder.cs (hunk 2 above)
Comment (AndriySvyryd) [inline src/EntityFramework.Core/Metadata/Internal/InternalEntityBuilder.cs]: It doesn't have to be any specific value, just a placeholder before the real PK is configured

File: src/EntityFramework.Core/Metadata/Property.cs (hunk 3 above)
Comment (AndriySvyryd) [inline src/EntityFramework.Core/Metadata/Property.cs]: Subjective

File: src/EntityFramework.Core/Metadata/Internal/InternalEntityBuilder.cs (hunk 1 above)
Comment (AndriySvyryd) [inline src/EntityFramework.Core/Metadata/Internal/InternalEntityBuilder.cs]: Looks better as is. The main purpose of the call is in the side effects, not the return value.

File: src/EntityFramework.Core/Metadata/Internal/InternalEntityBuilder.cs (hunk 2 above)
Comment (anpete) [inline src/EntityFramework.Core/Metadata/Internal/InternalEntityBuilder.cs]: "Placeholder"?

File: src/EntityFramework.Core/Metadata/Property.cs (hunk 3 above)
Comment (anpete) [inline src/EntityFramework.Core/Metadata/Property.cs]: I predicted that answer. Curious, do you do this for all x.Count == y.Count expressions? What is the criteria?

Comment (anpete) [inline src/EntityFramework.Core/Metadata/Property.cs]: Also "Equal" is more correct than "Same" here.

File: src/EntityFramework.Core/Metadata/Internal/InternalEntityBuilder.cs (hunk 2 above)
Comment (AndriySvyryd) [inline src/EntityFramework.Core/Metadata/Internal/InternalEntityBuilder.cs]: TempId?

File: src/EntityFramework.Core/Metadata/Property.cs (hunk 3 above)
Comment (AndriySvyryd) [inline src/EntityFramework.Core/Metadata/Property.cs]: It looks more consistent with ArePropertyTypesCompatible here.
I'll change the name to ArePropertyCountsEqual

**Comment by divega** [comment]
Blocked on #1741.

**Pull request merged**